keps/sig-cluster-lifecycle/kubeadm/4656-add-kubelet-instance-configuration/README.md

```diff
@@ -0,0 +1,287 @@
+
+## Motivation
+
+Currently, kubeadm adds a CRI socket annotation to the Node object during the `init upload-config` phase, which specifies the endpoint for the CRI that is being used by the kubelet on each node. This annotation is persistent on the Node object, even if the kubelet is updated or the CRI is changed.
+
+After migrating the container runtime endpoint flag to the instance configuration, we can use it
+Set the CRI socket by overriding the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml`.
+
+### Goals
+
+* kubeadm currently adds an annotation with the key `kubeadm.alpha.kubernetes.io/cri-socket` to each Node object. We will deprecate and remove it.
+* Provide an instance configuration file named `/var/lib/kubelet/instace-config.yaml` for each node, in which the `ContainerRuntimeEndpoint` field is defined. During the `kubeadm init/join/upgrade` process, the instance configuration will be read and the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` will be overwritten.
+* The `--container-runtime-endpoint` flag is no longer written to the `/var/lib/kubelet/kubeadm-flags.env` file
+
+### Non-Goals
+
+- Continue maintaining CRI socket paths on Node objects.
+
+## Proposal
+
+kubeadm init: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instance-config.yaml` configuration file based on it; if the CRI socket is not specified, the container runtime endpoint will be automatically detected, uploaded to the global configuration, and `/var/lib/kubelet/instace-config.yaml` will be generated.
+
+kubeadm join: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not specified, the kubelet configuration downloaded from the global configuration is used, and `/var/lib/kubelet/instace-config.yaml` is generated based on it.
+
+kubeadm upgrade:  future versions of `kubeadm upgrade apply/node` will only check `/var/lib/kubelet/instace-config.yaml`.
+
+### Risks and Mitigations
+
+## Design Details
+
+**Alpha(v1.31):** 
+
+kubeadm init:
+
+* No longer need to write the `--container-runtime-endpoint` to `/var/lib/kubelet/kubeadm-flags.env`.
+* No longer need to add the `kubeadm.alpha.kubernetes.io/cri-socket` annotation.
+* If the CRI socket provided in the kubeadm configuration is set, it is used first and the `/var/lib/kubelet/instace-config.yaml` configuration file is generated based on it. If the CRI socket is not set, the container runtime endpoint is automatically detected uploaded to the global configuration, and `/var/lib/kubelet/instace-config.yaml` will be generated.
```

Comment:
mistake instace->instance

check the whole doc for this problem please.

Reply:
> and `/var/lib/kubelet/instace-config.yaml` will be generated.

you mean that during init both the global config and the instance-config.yaml will have the auto-detected socket?

Reply:
> you mean that during init both the global config and the instance-config.yaml will have the auto-detected socket?

Yes, but I think this should be a mistake. We should not patch the CRI socket of any node to the global kubelet config.

Reply:
sorry, i still don't understand.
i think kubeadm should never use the socker that is provided in the global kubelet config that is in the config map?

today the annotation is always used unless the user passes a --config or --cri-socket that overrides it, correct?
the /var/lib/kubelet/config.yaml CRI socket field is new, but it's one of those things that don't make a lot of sense as kubeadm supports different container runtimes on different nodes.

Reply:
You are right, sorry for the misunderstanding, I meant the process I described in the KEP was wrong.